### PR TITLE
[AIRFLOW-3277] Correctly observe DST transitions for cron

### PR DESCRIPTION
**Pull request by bolkedebruin**
Make sure you have checked _all_ steps below.

### Jira

- [X] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-3277
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.

### Description

- [X] Here are some details about my PR, including screenshots of any UI changes:

`following_schedule` converts to naive time by using the
local time zone. In case of a DST transition, say 3AM -> 2AM
("summer time to winter time") we generate date times that
could overlap with earlier schedules. Therefore a DAG that
should run every 5 minutes will not do so if it has already
seen the schedule.

### Tests

- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

added

### Commits

- [X] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [X] In case of new functionality, my PR adds documentation that describes how to use it.
  - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.

### Code Quality

- [X] Passes `flake8`

cc @ashb

### CODE REVIEW

File: tests/models.py [hunk 1]
```diff
@@ -580,6 +580,29 @@ def test_cycle(self):
...
 
+    def test_following_previous_schedule(self):
+        """
+        Make sure DST transitions are properly observed
+        """
+        local_tz = pendulum.timezone('Europe/Zurich')
+        start = datetime(2018, 10, 28, 3, 00, tzinfo=local_tz)
+        utc = timezone.convert_to_utc(start)
+
+        dag = DAG('tz_dag', start_date=start, schedule_interval="'*/5 * * * *'")
```
Comment (ashb) [inline tests/models.py]:
```suggestion
        dag = DAG('tz_dag', start_date=start, schedule_interval='*/5 * * * *')
```

File: tests/models.py [hunk 2]
```diff
@@ -580,6 +580,29 @@ def test_cycle(self):
         with self.assertRaises(AirflowDagCycleException):
             dag.test_cycle()
 
+    def test_following_previous_schedule(self):
+        """
+        Make sure DST transitions are properly observed
+        """
+        local_tz = pendulum.timezone('Europe/Zurich')
+        start = datetime(2018, 10, 28, 3, 00, tzinfo=local_tz)
```
Comment (ashb) [inline tests/models.py]:
```suggestion
        start = datetime.datetime(2018, 10, 28, 3, 00, tzinfo=local_tz)
```

File: tests/models.py [hunk 3]
```diff
@@ -580,6 +580,29 @@ def test_cycle(self):
...
+        # should give 2018, 10, 28, 3, 5
+        self.assertEqual(next_local.hour, 3)
+        self.assertEqual(next_local.minute, 5)
+
+        prev = dag.previous_schedule(utc)
+        prev_local = local_tz.convert(prev)
+
+        # should give 2018, 10, 28, 1, 55
+        self.assertEqual(prev_local.hour, 1)
+        self.assertEqual(prev_local.minute, 55)
```
Comment (ashb) [inline tests/models.py]: I'm not quite sure we're constructing the DTs right here. (cos 2am local is ambigious time). This makes sure that the pre-condition time is correct:

```
        local_tz = pendulum.timezone('Europe/Zurich')
        start = local_tz.convert(datetime.datetime(2018, 10, 28, 2, 55), dst_rule=pendulum.PRE_TRANSITION)
        self.assertEqual(str(start), "2018-10-28 02:55:00+02:00", "Pre-condition: start date is in DST")

        utc = timezone.convert_to_utc(start)

        dag = DAG('tz_dag', start_date=start, schedule_interval='*/5 * * * *')
        next = dag.following_schedule(utc)
        next_local = local_tz.convert(next)

        self.assertEqual(str(next), "2018-10-28 01:00:00+00:00")
        self.assertEqual(str(next_local), "2018-10-28 02:00:00+01:00")

        prev = dag.previous_schedule(utc)
        prev_local = local_tz.convert(prev)

        self.assertEqual(str(prev_local), "2018-10-28 02:50:00+02:00")

        prev = dag.previous_schedule(next)
        prev_local = local_tz.convert(prev)

        self.assertEqual(str(prev_local), "2018-10-28 02:55:00+02:00")
        self.assertEqual(prev, utc)
```

Comment (ashb) [reply tests/models.py]: This change fixes it for hourly dags, but if (as was reported in Slack) you have a 3am daily dag I think this breaks the TZ feature:

```
>>> start = local_tz.convert(datetime.datetime(2018, 10, 27, 3, 0))
>>> utc = timezone.convert_to_utc(start)
>>> dag = DAG('tz_dag', start_date=start, schedule_interval='0 3 * * *')
>>> next = dag.following_schedule(utc)
>>> str(next)
2018-10-27 05:00:00+02:00
```

It is no longer 3am local time as requested :( (*edit*: fix the cron schedule field order in the example)

**Review by ashb — CHANGES_REQUESTED** [review]

File: tests/models.py (hunk 3 above)
Comment (ashb) [reply tests/models.py]: Tests that cover the 3am case:

```python
        start = local_tz.convert(datetime.datetime(2018, 10, 27, 3, 0))

        utc = timezone.convert_to_utc(start)

        dag = DAG('tz_dag', start_date=start, schedule_interval='0 3 * * *')
        next = dag.following_schedule(utc)
        next_local = local_tz.convert(next)

        self.assertEqual(str(next_local), "2018-10-28 03:00:00+01:00")
        self.assertEqual(str(next), "2018-10-28 02:00:00+00:00")

        prev = dag.previous_schedule(next)
        prev_local = local_tz.convert(prev)

        self.assertEqual(str(next_local), "2018-10-27 03:00:00+02:00") # This fails.
        self.assertEqual(str(next), "2018-10-27 01:00:00+00:00")
```

Buuut croniter's get_prev is flaky around DST transition :(

File: tests/models.py (hunk 1 above)
Comment (bolkedebruin) [reply tests/models.py]: Good point

File: tests/models.py (hunk 2 above)
Comment (bolkedebruin) [reply tests/models.py]: thanks

File: tests/models.py (hunk 3 above)
Comment (bolkedebruin) [reply tests/models.py]: Ah yes, this is why I did the initial conversion to local time with cron. Ie. the schedule is tied to local time, so if we use UTC and apply the schedule then we will make mistakes.

**Review by ashb — APPROVED** [review]

**Comment by rgangopadhya** [comment]
@cfei18 this seems to assume that the tzinfo has a name attribute, which is not always the case. 

```
>>> from dateutil import parser
>>> start_time = parser.parse("2018-11-28 00:00:00 -0800")
>> start_time.tzinfo.name
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'tzoffset' object has no attribute 'name'
```

Looks like timezones created with pendulum have a name, but not general timezones.

**Pull request merged**